File: src/array.cpp
#include "array.h"

#include <cstring>
#include <fstream>
#include <iostream>

using namespace std;

void DynamicArray::init(int initialCapacity) {
    data = new string[initialCapacity];
    size = 0;
    capacity = initialCapacity;
}

void DynamicArray::resize(int newCapacity) {
    string* newData = new string[newCapacity];
    for (int i = 0; i < size; i++) {
        newData[i] = data[i];
    }
    delete[] data;
    data = newData;
    capacity = newCapacity;
}

void DynamicArray::add(const string& value) {
    if (size == capacity) {
        resize(capacity * 2);
    }
    data[size++] = value;
}

void DynamicArray::insert(int index, const string& value) {
    if (index < 0 || index > size) {
        return;
    }
    if (size == capacity) {
        resize(capacity * 2);
    }
    for (int i = size; i > index; i--) {
        data[i] = data[i - 1];
    }
    data[index] = value;
    size++;
}

void DynamicArray::remove(int index) {
    if (index < 0 || index >= size) {
        return;
    }
    for (int i = index; i < size - 1; i++) {
        data[i] = data[i + 1];
    }
    size--;
}

string DynamicArray::get(int index) {
    if (index < 0 || index >= size) {
        return "";
    }
    return data[index];
}

void DynamicArray::set(int index, const string& value) {
    if (index < 0 || index >= size) {
        return;
    }
    data[index] = value;
}

int DynamicArray::length() { return size; }

void DynamicArray::print() {
    for (int i = 0; i < size; i++) {
        cout << data[i] << " ";
    }
    cout << endl;
}

void DynamicArray::destroy() { delete[] data; }

void DynamicArray::loadFromFile(const string& fileName) {
    ifstream file(fileName);
    string value;
    while (file >> value) {
        add(value);
    }
    file.close();
}

void DynamicArray::saveToFile(const string& fileName) {
    ofstream file(fileName);
    for (int i = 0; i < size; i++) {
        file << data[i] << endl;
    }
    file.close();
}

void runDynamicArray(int argc, char* argv[]) {
    DynamicArray arr;
    arr.init(10);

    string fileName;
    string query;

    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--file") == 0 && i + 1 < argc) {
            fileName = argv[i + 1];
            i++;
        } else if (strcmp(argv[i], "--query") == 0 && i + 1 < argc) {
            query = argv[i + 1];
            i++;
        }
    }

    arr.loadFromFile(fileName);

    string command;
    size_t pos = query.find(' ');
    if (pos != string::npos) {
        command = query.substr(0, pos);
        query = query.substr(pos + 1);
    } else {
        command = query;
    }

    if (command == "MPUSH") {
        arr.add(query);
        arr.saveToFile(fileName);
    } else if (command == "MINSERT") {
        size_t pos = query.find(' ');
        int index = stoi(query.substr(0, pos));
        string value = query.substr(pos + 1);
        arr.insert(index, value);
        arr.saveToFile(fileName);
    } else if (command == "MDEL") {
        int index = stoi(query);
        arr.remove(index);
        arr.saveToFile(fileName);
    } else if (command == "MSET") {
        size_t pos = query.find(' ');
        int index = stoi(query.substr(0, pos));
        string value = query.substr(pos + 1);
        arr.set(index, value);
        arr.saveToFile(fileName);
    } else if (command == "MLEN") {
        cout << arr.length() << endl;
    } else if (command == "PRINT") {
        arr.print();
    } else if (command == "MGET") {
        int index = stoi(query);
        cout << arr.get(index) << endl;
    }

    arr.destroy();
}

File: Go/src/array.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// DynamicArray — простая реализация динамического массива строк
type DynamicArray struct {
	data     []string
	capacity int
}

func (a *DynamicArray) Init(initialCapacity int) {
	if initialCapacity <= 0 {
		initialCapacity = 8
	}
	a.data = make([]string, 0, initialCapacity)
	a.capacity = initialCapacity
}

func (a *DynamicArray) resizeIfNeeded() {
	if len(a.data) == a.capacity {
		newCap := a.capacity * 2
		if newCap == 0 {
			newCap = 2
		}
		newData := make([]string, len(a.data), newCap)
		copy(newData, a.data)
		a.data = newData
		a.capacity = newCap
	}
}

func (a *DynamicArray) Add(value string) {
	a.resizeIfNeeded()
	a.data = append(a.data, value)
}

func (a *DynamicArray) Insert(index int, value string) {
	if index < 0 || index > len(a.data) {
		return
	}
	a.resizeIfNeeded()
	a.data = append(a.data, "")           // расширяем
	copy(a.data[index+1:], a.data[index:]) // сдвиг в право
	a.data[index] = value
}

func (a *DynamicArray) Remove(index int) {
	if index < 0 || index >= len(a.data) {
		return
	}
	copy(a.data[index:], a.data[index+1:])
	a.data = a.data[:len(a.data)-1]
}

func (a *DynamicArray) Get(index int) string {
	if index < 0 || index >= len(a.data) {
		return ""
	}
	return a.data[index]
}

func (a *DynamicArray) Set(index int, value string) {
	if index < 0 || index >= len(a.data) {
		return
	}
	a.data[index] = value
}

func (a *DynamicArray) Length() int {
	return len(a.data)
}

func (a *DynamicArray) Print() {
	for i := 0; i < len(a.data); i++ {
		if i > 0 {
			fmt.Print(" ")
		}
		fmt.Print(a.data[i])
	}
	fmt.Println()
}

func (a *DynamicArray) LoadFromFile(fileName string) error {
	file, err := os.Open(fileName)
	if err != nil {
		// если файла нет — оставляем пустой массив (как поведение C++: файл мог отсутствовать)
		return err
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	// сканируем по словам (по умолчанию Scanner.Split(bufio.ScanWords)), чтобы повторять поведение C++ >> operator
	for scanner.Scan() {
		a.Add(scanner.Text())
	}
	return scanner.Err()
}

func (a *DynamicArray) SaveToFile(fileName string) error {
	f, err := os.Create(fileName)
	if err != nil {
		return err
	}
	defer f.Close()
	w := bufio.NewWriter(f)
	for _, s := range a.data {
		_, _ = w.WriteString(s + "\n")
	}
	return w.Flush()
}

// runDynamicArray — выполняет команды для массива, интерфейс совместим с исходным.
func runDynamicArray(fileName string, fullQuery string) {
	arr := DynamicArray{}
	arr.Init(10)
	_ = arr.LoadFromFile(fileName) // игнорируем ошибку: возможно файл отсутствует

	// разделяем команду и аргументы
	parts := strings.SplitN(fullQuery, " ", 2)
	command := parts[0]
	arg := ""
	if len(parts) > 1 {
		arg = parts[1]
	}

	switch command {
	case "MPUSH":
		arr.Add(arg)
		if err := arr.SaveToFile(fileName); err != nil {
			fmt.Fprintln(os.Stderr, "Save error:", err)
		}
	case "MINSERT":
		// формат: "<index> <value>"
		sub := strings.SplitN(arg, " ", 2)
		if len(sub) != 2 {
			return
		}
		idx, err := strconv.Atoi(sub[0])
		if err != nil {
			return
		}
		arr.Insert(idx, sub[1])
		_ = arr.SaveToFile(fileName)
	case "MDEL":
		idx, err := strconv.Atoi(arg)
		if err != nil {
			return
		}
		arr.Remove(idx)
		_ = arr.SaveToFile(fileName)
	case "MSET":
		sub := strings.SplitN(arg, " ", 2)
		if len(sub) != 2 {
			return
		}
		idx, err := strconv.Atoi(sub[0])
		if err != nil {
			return
		}
		arr.Set(idx, sub[1])
		_ = arr.SaveToFile(fileName)
	case "MLEN":
		fmt.Println(arr.Length())
	case "MPRINT":
		arr.Print()
	case "MGET":
		idx, err := strconv.Atoi(arg)
		if err != nil {
			return
		}
		fmt.Println(arr.Get(idx))
	default:
		fmt.Fprintln(os.Stderr, "Неизвестная команда:", command)
	}
}
